### PR TITLE
Fixed issues with duplicate tests

### PR DESCRIPTION
**Pull request by samsinsane**
- Test suite will now fail when test names are reused
- Fixed several unit tests that reused test names

### CODE REVIEW

File: modules/self-test/test_declare.lua [hunk 1]
```diff
@@ -36,7 +36,14 @@
 			error('Duplicate test suite "'.. suiteName .. '"', 2)
 		end
 
-		local suite = {}
+		local _suite = {}
+		-- Setup a metatable for the test suites to use, this will catch duplicate tests
+		local suite = setmetatable({}, {
+			__index = _suite,
+			__newindex = function (table, key, value) if m.detectDuplicateTests and _suite[key] ~= nil then error('Duplicate test "'.. key .. '"', 2) end _suite[key] = value end,
```
Comment (tvandijck) [inline modules/self-test/test_declare.lua]: can we use normal multi-line code... I know it's just a styling thing, but it makes it a bit more readable.
also `p.error`.

**Comment by tvandijck** [comment]
could this be the dreaded random fail with just `next()` in the log ?

File: modules/self-test/test_declare.lua (hunk 1 above)
Comment (samsinsane) [reply modules/self-test/test_declare.lua]: Ah, my bad, I wasn't supposed to leave it like that.

Not sure why I used `error` over `p.error`, weird.

**Comment by samsinsane** [comment]
> could this be the dreaded random fail with just `next()` in the log ?

I've never seen this error before, I kind of doubt that it's related to this though. The duplicate tests just override the previous test, it's all in order so they always replace the same function. I would imagine that error is probably much more likely to be part of the alias fixes I made, there *were* 1-3 extra "tests" and 2 of them are set to `nil`. If, a big if here, `for testName, testFunction in pairs(test.suite) do` has an internal iterator, it's likely that the `next()` function fails because the table entry was removed when it was set to `nil`. No idea how the `pairs` function works, but that's my guess - that's assuming that either of my PRs resolves that `next()` issue.

File: modules/self-test/test_declare.lua (hunk 1 above)
Comment (samsinsane) [reply modules/self-test/test_declare.lua]: Oh, the top of this function uses `error`, that's why I used it. `p.error` doesn't provide the file location either, whereas `error` does, which is likely why `error` is used instead here.

Comment (samsinsane) [reply modules/self-test/test_declare.lua]: Example using `error`:
```
C:\github\premake-core>bin\release\premake5.exe test
Running action 'test'...
Error: ** Error: Duplicate test "referencesAreRelative_onDifferentProjectLocation"
```
Example using `p.error`:
```
C:\github\premake-core>bin\release\premake5.exe test
Running action 'test'...
Error: ...-core/modules/vstudio/tests/vc2010/test_project_refs.lua:83: Duplicate test "referencesAreRelative_onDifferentProjectLocation"
```

**Comment by samsinsane** [comment]
> Either way, this, and that alias test thing, are pretty darn good catches...

Yeah, I was trying to improve the output of the test runner and my numbers weren't lining up, so I sort of just stumbled across these two bugs. Potentially fixing that random fail sounds like a nice bonus! :+1:
